### PR TITLE
add asr2 task and librispeech recipe as an example. 

### PR DESCRIPTION
**Pull request by simpleoier**
This PR adds `discrete_asr1` task and librispeech recipe as an example. 
BPE dropout is enabled.

### CODE REVIEW

**Comment by sw005320** [comment]
@pengchengguo, can you review it?

@ftshijt, since it is a new template, could I also ask you to review this PR, especially for the library design level?

@simpleoier, can you consider a better name than `discrete_asr1`? How about `asr2` `dasr1`?

**Comment by pengchengguo** [comment]
Sure, I am happy to do it. Will finish it this week.

File: egs2/TEMPLATE/discrete_asr1/discrete_asr.sh
Comment (ftshijt) [inline egs2/TEMPLATE/discrete_asr1/discrete_asr.sh:350]: will token_joint to train bpe valid in this scenario?

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (ftshijt) [inline egs2/librispeech/discrete_asr1/local/data.sh:1]: I'm not sure making the whole data preprocessing to local is a good choice. My concern is that the method might need more extra efforts when applying to other datasets. For the first version, I think it might be good, but for generalization ability, at least we could organize some stages (kmeans etc.) to the pyscripts/scripts. My preference is still to put into TEMPLATE though, as that's the reason we hold the template for each task.

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (ftshijt) [inline egs2/librispeech/discrete_asr1/local/data.sh:188]: why not using the s3prl for ssl features, I believe that could be easier to maintain?

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (ftshijt) [inline egs2/librispeech/discrete_asr1/local/data.sh:225]: Do we have the distributed version for kmeans in espnet now?

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (ftshijt) [inline egs2/librispeech/discrete_asr1/local/data.sh:404]: Just curios, how you would handle the units with more than 5000 then?

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (ftshijt) [reply egs2/librispeech/discrete_asr1/local/data.sh:1]: I thought most of the functions are similar to hubert-pretraining one, can we reuse that?

**Review by ftshijt — COMMENTED** [review]
Thanks! very interesting design.

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/discrete_asr1/local/data.sh:225]: I think it still in progress. When we update it in the hubert recipe, it can be applied here in the same way.

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/discrete_asr1/local/data.sh:404]: Good catch. I should change the 4096 to nclusters, and adding a check if it is less than the total number of CJK tokens. I believe we can use up to 10K tokens.

**Comment by ftshijt** [comment]
Another suggestion, might be general: you may add a readme section probably as similar to
https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/tts1 

As most of our current tasks does not have that kinds of documentation, I recently found it very useful as a user manual (even better than the colab one as we hardly go through everything in the colab given its computation limitation). I've added that for svs and plan to do that for other tasks I know well: diar/uasr/asr.

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/discrete_asr1/local/data.sh]: I added s3prl support for this task, although I'm not sure if it matches the performance. 
But it is better to keep the script in ssl1 for better support of espnet-hubert models.

File: egs2/librispeech/discrete_asr1/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/discrete_asr1/local/data.sh]: Indeed. Now I reused kmeans training and label dumping from hubert-training.

File: egs2/TEMPLATE/asr2/conf/fbank.conf
Comment (Emrys365) [inline egs2/TEMPLATE/asr2/conf/fbank.conf:1]: Is this file still needed?

File: egs2/librispeech/asr2/conf/fbank.conf
Comment (Emrys365) [inline egs2/librispeech/asr2/conf/fbank.conf:1]: ditto

File: egs2/librispeech/asr2/conf/pitch.conf
Comment (Emrys365) [inline egs2/librispeech/asr2/conf/pitch.conf:1]: ditto

File: egs2/TEMPLATE/asr2/conf/pitch.conf
Comment (Emrys365) [inline egs2/TEMPLATE/asr2/conf/pitch.conf:1]: ditto

File: egs2/librispeech/asr2/local/data.sh
Comment (Emrys365) [inline egs2/librispeech/asr2/local/data.sh:71]: What if the input argument `skip_stages` is not empty and does not end with space?

File: egs2/TEMPLATE/asr2/conf/fbank.conf
Comment (simpleoier) [reply egs2/TEMPLATE/asr2/conf/fbank.conf:1]: Thanks for noticing it. I deleted it and the following unused files.

File: egs2/librispeech/asr2/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/asr2/local/data.sh]: add a space before `2`.

File: espnet2/tasks/mt.py
Comment (Emrys365) [inline espnet2/tasks/mt.py:70]: When will `specaug` be used?

File: espnet2/mt/frontend/embedding.py
Comment (Emrys365) [inline espnet2/mt/frontend/embedding.py:10]: Are they used?

File: espnet2/tasks/mt.py
Comment (simpleoier) [reply espnet2/tasks/mt.py:70]: [Here](https://github.com/espnet/espnet/pull/5181/files#diff-5f570b15a65d98e3e8992f846a74405495a848613069e73df8ecdf8157d1cad8R216)

File: espnet2/mt/frontend/embedding.py
Comment (simpleoier) [reply espnet2/mt/frontend/embedding.py:10]: Thanks for noticing it! I added by mistake.

File: espnet2/tasks/mt.py
Comment (Emrys365) [reply espnet2/tasks/mt.py:70]: If the input is already a discrete sequence, why would you still need the SpecAugment?

Comment (simpleoier) [reply espnet2/tasks/mt.py:70]: We verified that the time-masking is helping. Frequency-masking probably does not make sense. But the time-warping needs to be further verified. And the specaug is after the embedding layer.

File: egs2/librispeech/asr2/local/data.sh
Comment (sw005320) [inline egs2/librispeech/asr2/local/data.sh:1]: Since we create a new recipe instead of adjusting it to the existing `asr1` recipe, we may think of splitting the data preparation and discrete toke generation in the `asr2.sh` level instead of putting everything in the data preparation. Then, we can make the data generation part in a template instead of local processing.

The reason is that we may need to change some configurations (e.g., change the SSL models), and it would be easier to control by a top `asr2.sh` script instead of the local script.
Also, this framework seems to be solid and can be templated.

However, I also understand that it could be fully dependent on each recipe, and the current scheme might be better.

This is just a discussion point.

File: egs2/librispeech/asr2/local/dump_hubert_or_wavlm_feature.py
Comment (sw005320) [inline egs2/librispeech/asr2/local/dump_hubert_or_wavlm_feature.py:8]: ditto

File: egs2/mini_an4/ssl1/run.sh
Comment (sw005320) [inline egs2/mini_an4/ssl1/run.sh:17]: what is the difference between hubert and espnet_hubert?

File: espnet2/asr/pit_espnet_model.py
Comment (sw005320) [inline espnet2/asr/pit_espnet_model.py:94]: why did you use CPU here?

**Review by sw005320 — COMMENTED** [review]
LGTM.
After we confirm some points, we can merge this PR.

File: egs2/mini_an4/ssl1/run.sh
Comment (ftshijt) [reply egs2/mini_an4/ssl1/run.sh:17]: (hubert is from fairseq (so that we can use their checkpoint as well) while espnet_hubert is the one that from our side)

File: egs2/librispeech/asr2/local/data.sh
Comment (ftshijt) [reply egs2/librispeech/asr2/local/data.sh:1]: I vote for making it into template, as it would be more generalizable to other datasets (otherwise, the customization for  recipe users might be too much compared to other recipe designs).

One comment is that I also put the clustering into a stage as for S2ST, probably we could sync a bit to make it standard across different scenarios. As far as I know, four tasks (hubert ssl pretraining, discrete asr, s2st, and unsupervised asr) all use the same clustering extraction functions.

File: egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py
Comment (simpleoier) [reply egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py:11]: Yeah, I'll adjust this part.

File: egs2/librispeech/asr2/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/asr2/local/data.sh:1]: Sounds good. I was also thinking of moving things in the TEMPLATE. I'll do it.

File: egs2/mini_an4/ssl1/run.sh
Comment (simpleoier) [reply egs2/mini_an4/ssl1/run.sh:17]: So, in the previous script, we support the HuBERT ESPnet version, named as `hubert` as well as the legacy Fairseq version, named as `fairseq`. Since the `perform_kmeans.sh` inside the running script now supports `s3prl` based feature (including hubert as well). In this regard, I think renaming it as `fairseq_hubert` and `espnet_hubert` is more clear.

File: espnet2/asr/pit_espnet_model.py
Comment (simpleoier) [reply espnet2/asr/pit_espnet_model.py:94]: It's a bug fix. Maybe it shouldn't be in this PR. I found that `min_ids` is cuda tensor, and couldn't be used as the indices in slicing a tensor.

File: egs2/librispeech/asr2/local/data.sh
Comment (simpleoier) [reply egs2/librispeech/asr2/local/data.sh:1]: @sw005320 @ftshijt I think the new template can be divided into the following stages.
```
Stage 1 - 3  : the same as ASR1 template, including "local data preparation", "speed perturbation", "wav format"
Stage 4      : Kmeans script (scripts/feats/perform_kmeans.sh), including "ssl feature dump", "kmeans learning", "discrete token generation" and optionally label quality measurement.
Stage 5 - 6  : create dump folder and filter data
Stage 7      : Tokenization training
Stage 8 - 11 : LM related
Stage 12 - 15: ASR related (collect_stats, training, decoding, scoring)
Stage 16 - 18: Model packing and uploading.
```
Especially in stage 4, the `kmeans script` can be reused by SSL HuBERT without any change or other recipes.

File: espnet2/asr/pit_espnet_model.py
Comment (sw005320) [reply espnet2/asr/pit_espnet_model.py:94]: Can you leave the comments in the source code then?

File: egs2/librispeech/asr2/local/data.sh
Comment (ftshijt) [reply egs2/librispeech/asr2/local/data.sh:1]: Great to me as well! thanks for your effort

**Review by pengchengguo — APPROVED** [review]

**Review by pengchengguo — APPROVED** [review]

File: egs2/librispeech/ssl1/local/data.sh
Comment (simpleoier) [inline egs2/librispeech/ssl1/local/data.sh:41]: I changed the data.sh to reuse the stages from asr1. Similar changes were done for librispeech/ssl1/local/data.sh

File: egs2/TEMPLATE/asr2/asr2.sh
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:764]: I found a bug here and modified it.

**Review by simpleoier — COMMENTED** [review]
I resolved a bug in asr2.sh and modifed the librispeech/*/data.sh

File: egs2/librispeech/asr2/run.sh
Comment (sw005320) [inline egs2/librispeech/asr2/run.sh:1]: Can you try to remove the default arguments (and make some of them default and further remove them)?
Since we made a new template for this specific task, we can try to remove them so that it becomes a new standard recipe with minimized specifications.

File: egs2/mini_an4/asr2/run.sh
Comment (sw005320) [inline egs2/mini_an4/asr2/run.sh:1]: ditto
I think we can remove several arguments and make `run.sh` simple.

File: egs2/librispeech/asr2/run.sh
Comment (simpleoier) [reply egs2/librispeech/asr2/run.sh:1]: Yeah. How about this? I keep the following
```
./asr2.sh \
    --kmeans_feature "${kmeans_feature}" \
    --nclusters "${nclusters}" \
    --ngpu 2 \
    --nj 32 \
    --inference_nj 32 \
    --src_lang ${src_lang} \
    --tgt_lang ${tgt_lang} \
    --src_token_type "bpe" \
    --src_nbpe $src_nbpe \
    --tgt_token_type "bpe" \
    --tgt_nbpe $tgt_nbpe \
    --src_case ${src_case} \
    --tgt_case ${tgt_case} \
    --speed_perturb_factors "0.9 1.0 1.1" \
    --asr_config "${asr_config}" \
    --inference_config "${inference_config}" \
    --train_set "${train_set}" \
    --valid_set "${train_dev}" \
    --test_sets "${test_sets}" \
    --src_bpe_train_text "data/${train_set}/text.${src_case}.${src_lang}" \
    --tgt_bpe_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang}" \
    --lm_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang} data/local/other_text/text" "$@"
```
and remove
```
    --kmeans_opts "--stage 1 --stop-stage 4 --nj 4 --alignment data/mfa_phoneme_alignment" \
    --storage_save_mode true \
    --use_lm false \
    --token_joint false \
    --feats_type raw \
```

File: egs2/mini_an4/asr2/run.sh
Comment (sw005320) [reply egs2/mini_an4/asr2/run.sh:1]: Is the case information needed?

Comment (simpleoier) [reply egs2/mini_an4/asr2/run.sh:1]: Currently, the default `src_case` is "rm". However, the kmeans results for mini_an4 was too smooth. The resulting input sequence of deduplication was too short for subsampling layer in training. Probably the good way is that I can set the `src_case` to be `ts` by default to remove the the `*_case` arguments here.

File: egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py
Comment (pengchengguo) [reply egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py:11]: Is it possible to keep the original acknowledgment and then append the descriptions of your updates? I think we should acknowledge the original code and also point your updates.

File: egs2/TEMPLATE/asr2/setup.sh
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/setup.sh:44]: Here updated the script name from `asr.sh` to `asr2.sh`.

File: egs2/TEMPLATE/asr1/pyscripts/feats/dump_ssl_feature.py
Comment (Emrys365) [inline egs2/TEMPLATE/asr1/pyscripts/feats/dump_ssl_feature.py:67]: Which part of the SSL model is affected by the random seed?

File: egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py
Comment (simpleoier) [reply egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py:83]: It's used to unify the interface for all readers. Other readers need this argument.

File: egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh
Comment (Emrys365) [inline egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh:25]: This seems too specific to Librispeech.

Comment (simpleoier) [reply egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh:25]: Yeah, I intended to do so. Basically, the current quality measurement is specially designed for LibriSpeech. That's why the quality measure python script is under librispeech/{ssl1,asr2}/local. We need to update this part and the quality measurement script in the future. I'm afraid that the other dataset's alignment doesn't match the tsv format as used in librispeech.

File: egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh
Comment (Emrys365) [inline egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh:29]: These arguments are not clearly explained.

File: egs2/TEMPLATE/asr1/pyscripts/feats/dump_ssl_feature.py
Comment (simpleoier) [reply egs2/TEMPLATE/asr1/pyscripts/feats/dump_ssl_feature.py:67]: Good catch. I'll remove this one.

File: egs2/TEMPLATE/asr2/README.md
Comment (Emrys365) [inline egs2/TEMPLATE/asr2/README.md:37]: I think more information can be provided about this stage, because the used data files are different from the standard ones in Kaldi.

Comment (simpleoier) [reply egs2/TEMPLATE/asr2/README.md:37]: Can you be more specific? How different from standard ones is it?

File: egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh
Comment (simpleoier) [reply egs2/TEMPLATE/asr1/scripts/feats/perform_kmeans.sh]: I added a comment in the Stage 4.

File: egs2/TEMPLATE/asr2/README.md
Comment (Emrys365) [reply egs2/TEMPLATE/asr2/README.md:37]: For example, no `wav.scp` is used/generated. Some new files in addition to the standard files (wav.scp, text, utt2spk, spk2utt) are generated.

File: egs2/TEMPLATE/asr1/pyscripts/feats/dump_km_label.py
Comment (pengchengguo) [inline egs2/TEMPLATE/asr1/pyscripts/feats/dump_km_label.py:143]: Maybe we don't need to check whether `reader_conf["multilayer_feature"]` is a `str` or already `bool`, we can directly convert it to `bool`. This will make the code simpler and clearer.
Same to the following `reader_conf["layer"]`.

File: egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py
Comment (pengchengguo) [reply egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py:83]: Can we use `**kwargs` to refer all unsed params that for compatibility?

File: egs2/TEMPLATE/asr2/README.md
Comment (pengchengguo) [inline egs2/TEMPLATE/asr2/README.md:56]: Sorry, I am confused about the CJK characters. CJK commonly refers to Chinese, Japanese and Korean. Does it mean UTF-8 characters? Maybe I have misunderstood.

File: egs2/TEMPLATE/asr2/README.md
Comment (simpleoier) [reply egs2/TEMPLATE/asr2/README.md:37]: In recent commits, stage 1 generates wav.scp and used in stage 2 - 4. New files, to which I supposed you referred are the discrete token input text and the renamed target text, are generated in stage 4b now.

File: egs2/librispeech/asr2/run.sh
Comment (ftshijt) [inline egs2/librispeech/asr2/run.sh:35]:
```suggestion
```

**Review by ftshijt — APPROVED** [review]
LGTM! Thanks for the effort.

File: espnet2/train/preprocessor.py
Comment (pengchengguo) [inline espnet2/train/preprocessor.py:603]: For a normal ASR task that uses multiple tokenizer, like 3 different tokenizers (char, bpe2000, bpe5000), and the `tokenizer_encode_conf` is set to default, will the index `i` of `tokenizer_encode_conf` out of bound?

**Comment by sw005320** [comment]
@simpleoier, it’s almost there.
Can you reflect the other comments?
I think we can merged this PR soon.

**Comment by simpleoier** [comment]
Sure. I'll work on this today.

File: espnet2/train/preprocessor.py
Comment (simpleoier) [reply espnet2/train/preprocessor.py:603]: Yeah, I previously only considered the case for `src_tokenizer` and `tgt_tokenizer` and set the default value to be None. But thanks for pointing this out. I add another value check now.

File: egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py
Comment (simpleoier) [inline egs2/TEMPLATE/asr1/pyscripts/feats/ssl_feature_utils.py:82]: As @pengchengguo suggested, I used `kwargs` as a placeholder for unused arguments.

File: egs2/librispeech/asr2/run.sh
Comment (simpleoier) [inline egs2/librispeech/asr2/run.sh:33]: As @ftshijt suggested, I removed the `nj` and `inference_nj` which used the default values.

File: espnet2/train/preprocessor.py
Comment (simpleoier) [inline espnet2/train/preprocessor.py:603]: As @pengchengguo suggested, I add another value check here.

**Review by simpleoier — COMMENTED** [review]
Resolved the comments from some reviewers. Thanks for your suggestions!

File: egs2/TEMPLATE/asr2/asr2.sh
Comment (simpleoier) [inline egs2/TEMPLATE/asr2/asr2.sh:1200]: Update the log information: correct the stage number here

**Comment by sw005320** [comment]
I think this is a good shape.
I'll merge this after the CI check.

**Comment by simpleoier** [comment]
Thanks!

**Pull request merged**
